### PR TITLE
DMP-4475: Flyway Migrations for v72.2 Datamodel Update

### PR DESCRIPTION
**Pull request by chrisbellingham-hmcts**
## [DMP-4475](https://tools.hmcts.net/jira/browse/DMP-4475)

Flyway migrations to cover v72.2 datamodel updates per https://github.com/hmcts/darts-api/pull/2355.

Migrations have been validated against:
* a clean local database
* a local database pre-populated with staging data using `db-stg-to-local.sh`

Results from the latter:
```
2024-12-12T08:35:59.834 INFO  [main] org.flywaydb.core.internal.command.DbValidate Successfully validated 311 migrations (execution time 00:00.099s)
2024-12-12T08:35:59.850 INFO  [main] org.flywaydb.core.internal.command.DbMigrate Current version of schema "darts": 1.420
2024-12-12T08:35:59.863 INFO  [main] org.flywaydb.core.internal.command.DbMigrate Migrating schema "darts" to version "1.421 - v72 2"
2024-12-12T08:36:00.328 INFO  [main] o.f.c.internal.sqlscript.DefaultSqlScriptExecutor DB: index "usr_un_idx" does not exist, skipping
2024-12-12T08:36:00.379 INFO  [main] org.flywaydb.core.internal.command.DbMigrate Migrating schema "darts" to version "1.421.1 - v72 2"
2024-12-12T08:36:00.396 INFO  [main] org.flywaydb.core.internal.command.DbMigrate Successfully applied 2 migrations to schema "darts", now at version v1.421.1 (execution time 00:00.476s)
```

No other environments have been checked.

Additional to the migration scripts, changes include related:
* Entity updates
* Production and test code updates, to remove any usages of the soon-to-be-dropped `user_name` field

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/resources/db/migration/common/V1_421__v72_2.sql [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    created_by       INTEGER                  NOT NULL,
+    last_modified_ts TIMESTAMP WITH TIME ZONE NOT NULL,
+    last_modified_by INTEGER                  NOT NULL
+);
+
+ALTER TABLE extobjdir_process_detail
+    ADD CONSTRAINT epd_external_object_directory_fk
+        FOREIGN KEY (eod_id) REFERENCES external_object_directory (eod_id);
+
+-- TODO: FKs for created_by and last_modified_by
```
Comment (chrisbellingham-hmcts) [inline src/main/resources/db/migration/common/V1_421__v72_2.sql:78]: Author note: Awaiting comment resolution on https://github.com/hmcts/darts-api/pull/2355.

Comment (chrisbellingham-hmcts) [reply src/main/resources/db/migration/common/V1_421__v72_2.sql]: Resolving on the basis these FKs will be added in v72_3.

File: src/main/java/uk/gov/hmcts/darts/common/entity/UserAccountEntity.java [hunk 2]
```diff
@@ -82,6 +78,30 @@ public class UserAccountEntity extends CreatedModifiedBaseEntity {
...
+    @Column(name = "user_login_name")
+    private String userLoginName;
+
+    @NotAudited
+    @Column(name = "user_login_domain")
+    private String userLoginDomain;
+
+    @NotAudited
+    @Column(name = "user_state")
+    private Short userState;
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/darts/common/entity/UserAccountEntity.java:103]: What is the validation on this fields/

NotNull Max Lenght, Min Length, Regex checks etc?

File: src/main/java/uk/gov/hmcts/darts/common/entity/TranscriptionDocumentEntity.java [hunk 3]
```diff
@@ -97,4 +97,11 @@ public class TranscriptionDocumentEntity extends ModifiedBaseEntity
 
     @OneToMany(mappedBy = ObjectAdminActionEntity_.TRANSCRIPTION_DOCUMENT)
     private List<ObjectAdminActionEntity> adminActions = new ArrayList<>();
+
+    @Column(name = "subcontent_object_id")
+    private String subcontentObjectId;
+
+    @Column(name = "subcontent_position")
+    private Integer subcontentPosition;
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/darts/common/entity/TranscriptionDocumentEntity.java:105]: Validation (See previous comment)?

File: src/main/java/uk/gov/hmcts/darts/common/entity/SecurityGroupEntity.java [hunk 4]
```diff
@@ -93,4 +93,8 @@ public class SecurityGroupEntity extends MandatoryCreatedModifiedBaseEntity {
     @Column(name = "display_name")
     private String displayName;
 
+    @NotAudited
+    @Column(name = "group_display_name")
+    private String legacyGroupDisplayName;
+
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/darts/common/entity/SecurityGroupEntity.java:99]: Validation (See previous comment)?

File: src/main/java/uk/gov/hmcts/darts/common/entity/MediaEntity.java [hunk 5]
```diff
@@ -136,6 +136,11 @@ public class MediaEntity extends CreatedModifiedBaseEntity
     @OneToMany(mappedBy = ExternalObjectDirectoryEntity_.MEDIA)
     private List<ExternalObjectDirectoryEntity> externalObjectDirectoryEntities = new ArrayList<>();
 
+    @Column(name = "subcontent_object_id")
+    private String subcontentObjectId;
+
+    @Column(name = "subcontent_position")
+    private Integer subcontentPosition;
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/darts/common/entity/MediaEntity.java:143]: Validation (See previous comment)?

File: src/main/java/uk/gov/hmcts/darts/common/entity/ExternalObjectDirectoryProcessDetailEntity.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "eod_id")
+    private ExternalObjectDirectoryEntity externalObjectDirectory;
+
+    @Column(name = "event_date_ts")
+    private OffsetDateTime eventDateTime;
+
+    @Column(name = "update_retention")
+    private boolean updateRetention;
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/darts/common/entity/ExternalObjectDirectoryProcessDetailEntity.java:39]: Validation (See previous comment)?

File: src/main/java/uk/gov/hmcts/darts/common/entity/DailyListEntity.java [hunk 7]
```diff
@@ -107,4 +105,10 @@ public class DailyListEntity extends CreatedModifiedBaseEntity {
     @JoinColumn(name = "elt_id")
     private ExternalLocationTypeEntity externalLocationTypeEntity;
 
+    @Column(name = "subcontent_object_id")
+    private String subcontentObjectId;
+
+    @Column(name = "subcontent_position")
+    private Integer subcontentPosition;
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/darts/common/entity/DailyListEntity.java:112]: Validation (See previous comment)?

File: src/main/java/uk/gov/hmcts/darts/common/entity/CourthouseEntity.java [hunk 8]
```diff
@@ -70,10 +70,17 @@ public class CourthouseEntity extends CreatedModifiedBaseEntity {
...
     @Column(name = "display_name")
     private String displayName;
 
+    @NotAudited
+    @Column(name = "courthouse_object_id")
+    private String courthouseObjectId;
+
+    @NotAudited
+    @Column(name = "folder_path")
+    private String folderPath;
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/darts/common/entity/CourthouseEntity.java:82]: Validation (See previous comment)?

File: src/main/java/uk/gov/hmcts/darts/common/entity/AnnotationDocumentEntity.java [hunk 9]
```diff
@@ -91,6 +91,12 @@ public class AnnotationDocumentEntity extends ModifiedBaseEntity
     @Column(name = "ret_conf_reason")
     private String retConfReason;
 
+    @Column(name = "subcontent_object_id")
+    private String subcontentObjectId;
+
+    @Column(name = "subcontent_position")
+    private Integer subcontentPosition;
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/darts/common/entity/AnnotationDocumentEntity.java:98]: Validation (See previous comment)?

File: src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1_ok/expectedResponse.json [hunk 10]
```diff
@@ -20,5 +20,7 @@
   "contentObjectId": null,
   "clipId": null,
   "externalLocation": null,
-  "externalLocationTypeEntity": null
+  "externalLocationTypeEntity": null,
+  "subcontentObjectId": null,
+  "subcontentPosition": null
```
Comment (Ben-Edwards-cgi) [inline src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1_ok/expectedResponse.json:25]: Should this be in snake_case to follow HMCTs standards if they are new fields?

File: src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1_duplicate_ok/expectedResponse.json [hunk 11]
```diff
@@ -20,5 +20,7 @@
   "contentObjectId": null,
   "clipId": null,
   "externalLocation": null,
-  "externalLocationTypeEntity": null
+  "externalLocationTypeEntity": null,
+  "subcontentObjectId": null,
+  "subcontentPosition": null
```
Comment (Ben-Edwards-cgi) [inline src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1_duplicate_ok/expectedResponse.json:25]: Should this be in snake_case to follow HMCTs standards if they are new fields?

File: src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1OkJsonAndXml/expectedResponse2.json [hunk 12]
```diff
@@ -20,5 +20,7 @@
   "contentObjectId": null,
   "clipId": null,
   "externalLocation": null,
-  "externalLocationTypeEntity": null
+  "externalLocationTypeEntity": null,
+  "subcontentObjectId": null,
+  "subcontentPosition": null
```
Comment (Ben-Edwards-cgi) [inline src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1OkJsonAndXml/expectedResponse2.json:25]: Should this be in snake_case to follow HMCTs standards if they are new fields?

File: src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1OkJsonAndXml/expectedResponse.json [hunk 13]
```diff
@@ -20,5 +20,7 @@
   "contentObjectId": null,
   "clipId": null,
   "externalLocation": null,
-  "externalLocationTypeEntity": null
+  "externalLocationTypeEntity": null,
+  "subcontentObjectId": null,
+  "subcontentPosition": null
```
Comment (Ben-Edwards-cgi) [inline src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1OkJsonAndXml/expectedResponse.json:25]: Should this be in snake_case to follow HMCTs standards if they are new fields?

File: src/integrationTest/java/uk/gov/hmcts/darts/usermanagement/controller/UserControllerTest.java [hunk 14]
```diff
@@ -268,55 +268,6 @@ void testActivateModifyUserWithSuperAdmin() throws Exception {
         Assertions.assertEquals(workflowEntityBefore.size(), workflowEntityAfter.size());
     }
 
-    @Test
-    void testActivateModifyUserWithSuperAdminAndFailWithNoFullName() throws Exception {
```
Comment (Ben-Edwards-cgi) [inline src/integrationTest/java/uk/gov/hmcts/darts/usermanagement/controller/UserControllerTest.java:272]: Why was this removed? 
We still have fullName in the db?

File: src/main/java/uk/gov/hmcts/darts/common/entity/UserAccountEntity.java (hunk 2 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/common/entity/UserAccountEntity.java:103]: I think it makes it easier to identify what validation is applied to the entity so that when you are scripting you can really easily / quickly identify what a valid value should look like, without having to go to the data model, and it helps with belt and braces as we enforce the restriction across the board

I do agree with the inconsistencies that can bring so I will leave this up to your discretion :)

File: src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1_ok/expectedResponse.json (hunk 10 above)
Comment (chrisbellingham-hmcts) [reply src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1_ok/expectedResponse.json:25]: So this isn't actually an API response (despite the name of the file). Some tests take the approach of using a JSON file to represent the expected state of an object, and then serialising the actual object and comparing it to the expected JSON with `JSONAssert`. I'm not a fan of the approach personally.

So here, all of the attribute names actually correspond to the attribute names of a `DailyListEntity`.

File: src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1_duplicate_ok/expectedResponse.json (hunk 11 above)
Comment (chrisbellingham-hmcts) [reply src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1_duplicate_ok/expectedResponse.json:25]: See response to similar comment.

File: src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1OkJsonAndXml/expectedResponse2.json (hunk 12 above)
Comment (chrisbellingham-hmcts) [reply src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1OkJsonAndXml/expectedResponse2.json:25]: See response to similar comment.

File: src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1OkJsonAndXml/expectedResponse.json (hunk 13 above)
Comment (chrisbellingham-hmcts) [reply src/integrationTest/resources/tests/dailylist/DailyListServiceTest/insert1OkJsonAndXml/expectedResponse.json:25]: See response to similar comment.

File: src/integrationTest/java/uk/gov/hmcts/darts/usermanagement/controller/UserControllerTest.java (hunk 14 above)
Comment (chrisbellingham-hmcts) [reply src/integrationTest/java/uk/gov/hmcts/darts/usermanagement/controller/UserControllerTest.java:272]: The test case involves an attempt to activate a `user_account` who has a null `user_full_name`. The changes in this PR add a `NOT NULL` constraint to that field, so this is no longer a possible scenario.

This comment has prompted me to revise the `UserActivateValidator` class which checked for this condition.

File: src/main/java/uk/gov/hmcts/darts/common/entity/UserAccountEntity.java (hunk 2 above)
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/common/entity/UserAccountEntity.java:103]: On the basis the datamodel is now mature and further changes should (!) be minimal, I have now included the data layer constraints on the Entities. These are limited to `nullable = false`, and `length = n` - we don't know the full shape of the data so I'd rather not overconstrain with min length or regex checks.

File: src/main/java/uk/gov/hmcts/darts/common/entity/TranscriptionLinkedCaseEntity.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "cas_id")
+    private CourtCaseEntity courtCase;
+
+    @Column(name = "courthouse_name")
+    private String courthouseName;
+
+    @Column(name = "case_number")
+    private String caseNumber;
```
[the remark this answers is not in the document]
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/common/entity/TranscriptionLinkedCaseEntity.java]: See earlier response to similar comment.

File: src/main/java/uk/gov/hmcts/darts/common/entity/TranscriptionDocumentEntity.java (hunk 3 above)
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/common/entity/TranscriptionDocumentEntity.java]: See earlier response to similar comment.

File: src/main/java/uk/gov/hmcts/darts/common/entity/SecurityGroupEntity.java (hunk 4 above)
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/common/entity/SecurityGroupEntity.java:99]: See earlier response to similar comment.

File: src/main/java/uk/gov/hmcts/darts/common/entity/MediaEntity.java (hunk 5 above)
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/common/entity/MediaEntity.java]: See earlier response to similar comment.

File: src/main/java/uk/gov/hmcts/darts/common/entity/ExternalObjectDirectoryProcessDetailEntity.java (hunk 6 above)
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/common/entity/ExternalObjectDirectoryProcessDetailEntity.java]: See earlier response to similar comment.

File: src/main/java/uk/gov/hmcts/darts/common/entity/DailyListEntity.java (hunk 7 above)
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/common/entity/DailyListEntity.java]: See earlier response to similar comment.

File: src/main/java/uk/gov/hmcts/darts/common/entity/AnnotationDocumentEntity.java (hunk 9 above)
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/common/entity/AnnotationDocumentEntity.java]: See earlier response to similar comment.

**Review by Ben-Edwards-cgi — APPROVED** [review]
